src/error.ts

```diff
@@ -0,0 +1,47 @@
+interface FirebaseErrorOptions {
+  children?: unknown[];
+  context?: unknown;
+  exit?: number;
+  original?: Error;
+  status?: number;
+}
+
+const DEFAULT_CHILDREN: NonNullable<FirebaseErrorOptions["children"]> = [];
+const DEFAULT_EXIT: NonNullable<FirebaseErrorOptions["exit"]> = 1;
+const DEFAULT_STATUS: NonNullable<FirebaseErrorOptions["status"]> = 500;
+
+export class FirebaseError extends Error implements FirebaseErrorOptions {
+  readonly children: unknown[];
+
+  readonly context: unknown | undefined;
+
+  readonly exit: number;
+
+  readonly message: string;
+
+  readonly name = "FirebaseError";
+
+  readonly original: Error | undefined;
+
+  readonly status: number;
+
+  constructor(
+    message: string,
+    {
+      children = DEFAULT_CHILDREN,
+      context,
+      exit = DEFAULT_EXIT,
+      original,
+      status = DEFAULT_STATUS,
+    }: FirebaseErrorOptions = {}
```

Comment:
I'm not a huge fan of this large destructuring/defaulting pattern. Once an object gets to a certain point, destructuring adds more noise and complexity for the reader (this is the _third_ place that all the FirebaseErrorOptions are listed in this file...)

Could we keep it simple and just use `options` as a variable (with a default of an empty object, of course).

```typescript
  constructor(message: string, options: FirebaseErrorOptions = {}) {
    super();

    this.children = options.children || DEFAULT_CHILDREN;
    this.context = options.context;
    this.exit = options.exit || DEFAULT_EXIT;
    this.message = message;
    this.original = options.original;
    this.status = options.status || DEFAULT_STATUS;
  }
```

Reply:
5868765bee62fb0e5f365865a708f325fbc3240a